cmd/firewall.go

```diff
@@ -173,3 +202,238 @@ func (c *firewallCmd) createRequestFromCLI() (*models.V1FirewallCreateRequest, e
 		Ips:                mcr.Ips,
 	}, nil
 }
+
+func (c *config) firewallSSH(args []string) (err error) {
+	if len(args) < 1 {
+		return fmt.Errorf("machine ID is expected as an argument")
+	}
+	firewallID := args[0]
+	resp, err := c.client.Firewall().FindFirewall(firewall.NewFindFirewallParams().WithID(firewallID), nil)
+	if err != nil {
+		return fmt.Errorf("failed to find firewall: %w", err)
+	}
+
+	if resp.Payload.Vpn != nil && resp.Payload.Vpn.Connected != nil && *resp.Payload.Vpn.Connected {
+		return c.firewallSSHViaVPN(args)
+	}
+
+	// Try to connect to firewall via SSH
+	if err := c.firewallPureSSH(resp.Payload.Allocation); err != nil {
+		return fmt.Errorf("failed to connect to firewall via SSH: %w", err)
+	}
+
+	return nil
+}
+
+func (c *config) firewallPureSSH(fwAllocation *models.V1MachineAllocation) (err error) {
+	networks := fwAllocation.Networks
+	for _, nw := range networks {
+		if *nw.Underlay || *nw.Private {
+			continue
+		}
+		for _, ip := range nw.Ips {
+			if portOpen(ip, "22", time.Second) {
+				err = SSHClient("metal", viper.GetString("identity"), ip, 22)
+				if err != nil {
+					return fmt.Errorf("machine console error:%w", err)
+				}
+
+				return nil
+			}
+		}
+	}
+
+	return fmt.Errorf("no ip with a open ssh port found")
+}
+
+func (c *config) firewallSSHViaVPN(args []string) (err error) {
+	if runtime.GOOS != "linux" {
+		return fmt.Errorf("firewall ssh command isn't supported by your OS(only Linux support at the moment")
+	}
+	if _, err := os.Stat("/dev/net/tun"); errors.Is(err, os.ErrNotExist) {
+		return fmt.Errorf("/dev/net/tun file is missing, tun module needs to be loaded")
+	}
+
+	if len(args) < 1 {
+		return fmt.Errorf("machine ID is expected as an argument")
+	}
+	firewallID := args[0]
+
+	// Get firewall's project
+	machineGetResp, err := c.client.Machine().FindMachine(machine.NewFindMachineParams().WithID(firewallID), nil)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve firewall details: %w", err)
+	}
+	projectID := machineGetResp.Payload.Allocation.Project
+
+	isEphemeral := true
+	authKeyResp, err := c.client.VPN().GetVPNAuthKey(vpn.NewGetVPNAuthKeyParams().WithBody(&models.V1VPNRequest{
+		Pid:       projectID,
+		Ephemeral: &isEphemeral,
+	}), nil)
+	if err != nil {
+		return fmt.Errorf("failed to get VPN auth key: %w", err)
+	}
+
+	cli, err := dockerclient.NewClientWithOpts(dockerclient.FromEnv)
+	if err != nil {
+		return fmt.Errorf("failed to initialize Docker client: %w", err)
+	}
+
+	// Deploy tailscaled
+	ctx := context.Background()
+	if err := pullImageIfNotExists(ctx, cli, tailscaleImage); err != nil {
+		return fmt.Errorf("failed to pull tailscale image: %w", err)
+	}
+
+	containerConfig := &container.Config{
+		Image: tailscaleImage,
+		Cmd:   []string{"tailscaled"},
+	}
+	hostConfig := &container.HostConfig{
+		Binds: []string{
+			"/dev/net/tun:/dev/net/tun",
```

Comment:
Maybe we can also get rid of this by using userspace networking for tailscale as described here:
https://tailscale.com/kb/1112/userspace-networking/

Reply:
Done